File: harness/acs/acs1.py
"""ACS-1 — Alfred Canonical Serialization, version 1 (ADR-0003, ADR-0004).

The canonical byte form of any structure whose hash is load-bearing: evidence-chain
rows and result stamps. Artifacts are hashed as stored bytes and never pass through
here.

The whole design exists so that a third party — an auditor recomputing the evidence
chain, in a language that is not Python — produces byte-identical output. Every rule
below is therefore stated as an algorithm with no appeal to a host language's
formatting defaults.

No third-party imports, deliberately: the encoder that guards the audit chain must not
depend on the supply chain the chain records.
"""

from __future__ import annotations

import hashlib
import json
import unicodedata
from decimal import Decimal
from typing import Any

ACS_VERSION = "ACS-1"

INT64_MIN = -(2**63)
INT64_MAX = 2**63 - 1

# Two-character escapes JSON defines. Every other control character below 0x20 is
# emitted as \u00xx with lowercase hex. Nothing else is ever escaped — not 0x7f,
# not '/', not any non-ASCII character.
_SHORT_ESCAPES = {
    0x08: "\\b",
    0x09: "\\t",
    0x0A: "\\n",
    0x0C: "\\f",
    0x0D: "\\r",
    0x22: '\\"',
    0x5C: "\\\\",
}


class AcsError(ValueError):
    """Base for every refusal. Each carries a stable code; the codes are part of the
    published test-vector suite, so an independent implementation can be checked on
    what it rejects and not merely on what it emits."""

    code = "ACS_ERROR"


class NotFinite(AcsError):
    code = "NOT_FINITE"


class IntOutOfRange(AcsError):
    code = "INT_OUT_OF_RANGE"


class DuplicateKey(AcsError):
    code = "DUPLICATE_KEY"


class BadKeyType(AcsError):
    code = "BAD_KEY_TYPE"


class UnsupportedType(AcsError):
    code = "UNSUPPORTED_TYPE"


class LoneSurrogate(AcsError):
    code = "LONE_SURROGATE"


# --------------------------------------------------------------------------- floats


def encode_float(value: float) -> str:
    """ACS-1 float grammar: normalized scientific, shortest round-tripping digits.

        sign? digit "." digit+ "e" "-"? exponent

    Always exactly one digit before the point, always at least one after, exponent
    with no leading zeros and no '+' sign. `1.0` is `"1.0e0"`, `100.0` is `"1.0e2"`,
    `2.4` is `"2.4e0"`, `1e-7` is `"1.0e-7"`.

    The presentation is pinned because "shortest round-trip" alone is not a
    specification: every correct implementation agrees on the *digits* and disagrees
    on the *formatting*. Python renders 1.0 as "1.0" and 1e16 as "1e+16"; JavaScript
    renders them "1" and "10000000000000000". Left to a host language's default, the
    string encoding would reintroduce the divergence it exists to remove.

    NaN and infinity are refused. Infinity is representable only through the tagged
    form of ADR-0001, and NaN is forbidden as an output anywhere in the system.
    """
    if value != value or value in (float("inf"), float("-inf")):
        raise NotFinite(f"float is not finite: {value!r}")

    if value == 0.0:
        return "0.0e0"  # also the normalization of -0.0

    sign, digits, exponent = Decimal(repr(value)).as_tuple()
    assert isinstance(exponent, int)

    # value == digits * 10**exponent, so in d.ddd * 10**e form:
    exp = (len(digits) - 1) + exponent

    # repr may carry trailing zeros that the shortest-digit form does not
    # (repr(100.0) == "100.0" -> digits (1,0,0,0)). Stripping them leaves `exp`
    # unchanged: dropping k trailing zeros shortens the digit string by k and raises
    # the decimal exponent by k, and exp is their sum.
    ds = "".join(str(d) for d in digits).rstrip("0") or "0"

    mantissa = ds[0] + "." + (ds[1:] or "0")
    return f"{'-' if sign else ''}{mantissa}e{exp}"


def decode_float(text: str) -> float:
    """Inverse of `encode_float`. Exact for every finite double."""
    return float(text)


# -------------------------------------------------------------------------- strings


def _normalize(text: str) -> str:
    """NFC. Composed and decomposed forms of the same string are visually identical
    and byte-different; without normalization they would hash differently."""
    try:
        text.encode("utf-8")
    except UnicodeEncodeError as exc:  # lone surrogate
        raise LoneSurrogate(f"string is not encodable as UTF-8: {exc}") from exc
    return unicodedata.normalize("NFC", text)


def _encode_string(text: str) -> str:
    out = ['"']
    for ch in text:
        cp = ord(ch)
        if cp in _SHORT_ESCAPES:
            out.append(_SHORT_ESCAPES[cp])
        elif cp < 0x20:
            out.append(f"\\u{cp:04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


# --------------------------------------------------------------------------- encode


def _encode(value: Any) -> str:
    # bool before int: in Python `isinstance(True, int)` is true, so an unguarded
    # int branch would silently serialize True as 1 and collapse two distinct values.
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, int):
        if not INT64_MIN <= value <= INT64_MAX:
            raise IntOutOfRange(f"integer outside signed 64-bit range: {value}")
        return str(value)
    if isinstance(value, float):
        return _encode_string(encode_float(value))
    if isinstance(value, str):
        return _encode_string(_normalize(value))
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(_encode(v) for v in value) + "]"
    if isinstance(value, dict):
        items: list[tuple[bytes, str, Any]] = []
        seen: set[bytes] = set()
        for raw_key, val in value.items():
            if not isinstance(raw_key, str):
                raise BadKeyType(f"object key is not a string: {raw_key!r}")
            key = _normalize(raw_key)
            kb = key.encode("utf-8")
            if kb in seen:
                # After normalization, because two keys differing only by composition
                # are the same key and must not both survive into the canonical form.
                raise DuplicateKey(f"duplicate object key after NFC: {key!r}")
            seen.add(kb)
            items.append((kb, key, val))
        items.sort(key=lambda it: it[0])
        body = ",".join(f"{_encode_string(k)}:{_encode(v)}" for _, k, v in items)
        return "{" + body + "}"
    raise UnsupportedType(f"type is not representable in ACS-1: {type(value).__name__}")


def canonicalize(value: Any) -> bytes:
    """The canonical UTF-8 bytes of `value`. No whitespace anywhere."""
    return _encode(value).encode("utf-8")


# ----------------------------------------------------------------------------- hash


def hash_input(record_type: str, value: Any) -> bytes:
    """Domain-separated preimage.

        ACS_VERSION 0x00 record_type 0x00 canonical_bytes

    The separators exist so a result stamp and an evidence row with coincidentally
    identical content cannot produce the same digest, and so the scheme version can
    change without silently reinterpreting existing digests.
    """
    if "\x00" in record_type:
        raise AcsError("record_type must not contain a NUL byte")
    return (
        ACS_VERSION.encode("utf-8")
        + b"\x00"
        + _normalize(record_type).encode("utf-8")
        + b"\x00"
        + canonicalize(value)
    )


def acs_sha256(record_type: str, value: Any) -> str:
    return hashlib.sha256(hash_input(record_type, value)).hexdigest()


# ---------------------------------------------------------------------------- parse


def _no_duplicates(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    keys = [_normalize(k) for k, _ in pairs]
    if len(set(keys)) != len(keys):
        raise DuplicateKey(f"duplicate object key: {keys}")
    return dict(zip(keys, (v for _, v in pairs)))


def _reject_constant(token: str) -> Any:
    raise NotFinite(f"JSON constant not permitted in ACS-1: {token}")


def _reject_float(token: str) -> Any:
    # Canonical form carries floats as strings, so a JSON float literal cannot occur in
    # it. Accepting one here would let `parse_strict` succeed on input the encoder can
    # never produce.
    raise AcsError(f"ACS-1 numbers are integers; floats are carried as strings: {token}")


def _check_int(token: str) -> int:
    value = int(token)
    if not INT64_MIN <= value <= INT64_MAX:
        raise IntOutOfRange(f"integer outside signed 64-bit range: {token}")
    return value


def parse_strict(data: bytes) -> Any:
    """Parse ACS-1 bytes, refusing what the stock parser accepts silently: duplicate
    keys (which the default keeps last-wins with no error) and the non-standard
    `NaN` / `Infinity` literals.

    Floats come back as strings. Callers know from their schema which fields are
    floats; the canonical form deliberately does not carry that distinction, because
    a self-describing numeric type is what created the divergence in the first place.
    """
    return json.loads(
        data.decode("utf-8"),
        object_pairs_hook=_no_duplicates,
        parse_constant=_reject_constant,
        parse_float=_reject_float,
        parse_int=_check_int,
    )


def is_canonical(data: bytes) -> bool:
    """Whether `data` is already in canonical form — the check an auditor runs on a
    stored row before trusting its digest."""
    try:
        return canonicalize(parse_strict(data)) == data
    except (AcsError, ValueError, UnicodeDecodeError):
        return False

File: harness/acs/acs1.mjs
// ACS-1 — independent JavaScript implementation (ADR-0003, ADR-0004).
//
// Written from the specification and the vector file, not translated from acs1.py.
// Its purpose is to test the claim the whole scheme rests on: that a third party can
// recompute Alfred's evidence-chain digests without running Alfred's code.
//
// An auditor needs the parser more than the encoder. They receive stored rows they did
// not produce, so `isCanonical` — parse, re-encode, compare — is the operation that
// actually answers "is this row what its digest says it is".
//
// No dependencies beyond node: built-ins.

import { createHash } from "node:crypto";

export const ACS_VERSION = "ACS-1";

export const INT64_MIN = -(2n ** 63n);
export const INT64_MAX = 2n ** 63n - 1n;

export class AcsError extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
  }
}

// JavaScript cannot tell 1 from 1.0 — both are Number — so the int/float distinction
// that ACS-1 needs on the way *in* cannot be recovered from the value. Floats are
// therefore wrapped explicitly and integers must be BigInt. A bare Number is refused
// rather than guessed at.
export class F64 {
  constructor(value) {
    this.value = value;
  }
}
export const f64 = (value) => new F64(value);

const SHORT_ESCAPES = new Map([
  [0x08, "\\b"],
  [0x09, "\\t"],
  [0x0a, "\\n"],
  [0x0c, "\\f"],
  [0x0d, "\\r"],
  [0x22, '\\"'],
  [0x5c, "\\\\"],
]);

// ------------------------------------------------------------------------- floats

export function encodeFloat(x) {
  if (typeof x !== "number") throw new AcsError("UNSUPPORTED_TYPE", `not a number: ${x}`);
  if (!Number.isFinite(x)) throw new AcsError("NOT_FINITE", `float is not finite: ${x}`);
  if (x === 0) return "0.0e0"; // also normalizes -0

  // toExponential() with no argument is specified to emit as many significand digits
  // as are needed to uniquely identify the value — the same shortest-digit sequence
  // every correct implementation produces. Only the presentation differs, and
  // ADR-0004 pins that.
  const [mantissa, exponent] = x.toExponential().split("e");
  const dot = mantissa.indexOf(".");
  const head = dot === -1 ? mantissa : mantissa.slice(0, dot);
  const tail = dot === -1 ? "0" : mantissa.slice(dot + 1);
  return `${head}.${tail}e${parseInt(exponent, 10)}`;
}

export const decodeFloat = (text) => Number(text);

// ------------------------------------------------------------------------ strings

function checkSurrogates(text) {
  // TextEncoder replaces a lone surrogate with U+FFFD and reports nothing, so the
  // corruption is silent unless it is checked for explicitly. Python raises here.
  for (let i = 0; i < text.length; i++) {
    const c = text.charCodeAt(i);
    if (c >= 0xd800 && c <= 0xdbff) {
      const next = text.charCodeAt(i + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new AcsError("LONE_SURROGATE", `unpaired high surrogate at ${i}`);
      }
      i++;
    } else if (c >= 0xdc00 && c <= 0xdfff) {
      throw new AcsError("LONE_SURROGATE", `unpaired low surrogate at ${i}`);
    }
  }
}

function normalize(text) {
  checkSurrogates(text);
  return text.normalize("NFC");
}

function encodeString(text) {
  let out = '"';
  for (const ch of text) {
    const cp = ch.codePointAt(0);
    if (SHORT_ESCAPES.has(cp)) out += SHORT_ESCAPES.get(cp);
    else if (cp < 0x20) out += "\\u" + cp.toString(16).padStart(4, "0");
    else out += ch;
  }
  return out + '"';
}

// ------------------------------------------------------------------------- encode

const UTF8 = new TextEncoder();

function compareBytes(a, b) {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) if (a[i] !== b[i]) return a[i] - b[i];
  return a.length - b.length;
}

function encode(value) {
  if (value === null) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "bigint") {
    if (value < INT64_MIN || value > INT64_MAX) {
      throw new AcsError("INT_OUT_OF_RANGE", `outside signed 64-bit range: ${value}`);
    }
    return value.toString();
  }
  if (value instanceof F64) return encodeString(encodeFloat(value.value));
  if (typeof value === "number") {
    throw new AcsError(
      "UNSUPPORTED_TYPE",
      "bare Number is ambiguous in ACS-1: use BigInt for integers, f64() for floats",
    );
  }
  if (typeof value === "string") return encodeString(normalize(value));
  if (Array.isArray(value)) return "[" + value.map(encode).join(",") + "]";

  const entries = value instanceof Map ? [...value.entries()] : null;
  if (entries === null) {
    throw new AcsError("UNSUPPORTED_TYPE", `not representable: ${Object.prototype.toString.call(value)}`);
  }

  const items = [];
  const seen = new Set();
  for (const [rawKey, val] of entries) {
    if (typeof rawKey !== "string") {
      throw new AcsError("BAD_KEY_TYPE", `object key is not a string: ${rawKey}`);
    }
    const key = normalize(rawKey);
    const bytes = UTF8.encode(key);
    const marker = bytes.join(",");
    if (seen.has(marker)) {
      throw new AcsError("DUPLICATE_KEY", `duplicate object key after NFC: ${key}`);
    }
    seen.add(marker);
    items.push([bytes, key, val]);
  }
  items.sort((a, b) => compareBytes(a[0], b[0]));
  return "{" + items.map(([, k, v]) => `${encodeString(k)}:${encode(v)}`).join(",") + "}";
}

export const canonicalize = (value) => UTF8.encode(encode(value));

// --------------------------------------------------------------------------- hash

export function hashInput(recordType, value) {
  if (recordType.includes("\u0000")) {
    throw new AcsError("ACS_ERROR", "record_type must not contain a NUL byte");
  }
  return Buffer.concat([
    Buffer.from(ACS_VERSION, "utf8"),
    Buffer.from([0]),
    Buffer.from(normalize(recordType), "utf8"),
    Buffer.from([0]),
    Buffer.from(canonicalize(value)),
  ]);
}

export const acsSha256 = (recordType, value) =>
  createHash("sha256").update(hashInput(recordType, value)).digest("hex");

// -------------------------------------------------------------------------- parse

// A hand-written parser rather than JSON.parse, for the same reason Python needs
// object_pairs_hook: the built-in silently keeps the last of a duplicate key pair, so
// the tampering it would hide is invisible to a reviver.
export function parseStrict(bytes) {
  const text = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  let i = 0;

  const fail = (code, msg) => {
    throw new AcsError(code, `${msg} at offset ${i}`);
  };
  const expect = (ch) => {
    if (text[i] !== ch) fail("PARSE_ERROR", `expected '${ch}'`);
    i++;
  };

  function parseValue() {
    const ch = text[i];
    if (ch === "{") return parseObject();
    if (ch === "[") return parseArray();
    if (ch === '"') return parseString();
    if (text.startsWith("true", i)) return (i += 4), true;
    if (text.startsWith("false", i)) return (i += 5), false;
    if (text.startsWith("null", i)) return (i += 4), null;
    if (text.startsWith("NaN", i) || text.startsWith("Infinity", i) || text.startsWith("-Infinity", i)) {
      fail("NOT_FINITE", "JSON constant not permitted in ACS-1");
    }
    return parseNumber();
  }

  function parseObject() {
    expect("{");
    const map = new Map();
    if (text[i] === "}") return i++, map;
    for (;;) {
      // Normalized before the duplicate check, and stored normalized, because two
      // keys differing only by composition are one key — the same rule the encoder
      // applies. Without this the parser accepts a row the encoder could never have
      // produced, which is exactly the tampering `parseStrict` exists to catch.
      const key = normalize(parseString());
      if (map.has(key)) fail("DUPLICATE_KEY", `duplicate object key '${key}'`);
      expect(":");
      map.set(key, parseValue());
      if (text[i] === ",") { i++; continue; }
      expect("}");
      return map;
    }
  }

  function parseArray() {
    expect("[");
    const out = [];
    if (text[i] === "]") return i++, out;
    for (;;) {
      out.push(parseValue());
      if (text[i] === ",") { i++; continue; }
      expect("]");
      return out;
    }
  }

  function parseString() {
    expect('"');
    let out = "";
    for (;;) {
      const ch = text[i];
      if (ch === undefined) fail("PARSE_ERROR", "unterminated string");
      if (ch === '"') return i++, out;
      if (ch === "\\") {
        i++;
        const esc = text[i++];
        const simple = { '"': '"', "\\": "\\", "/": "/", b: "\b", f: "\f", n: "\n", r: "\r", t: "\t" };
        if (esc in simple) out += simple[esc];
        else if (esc === "u") {
          out += String.fromCharCode(parseInt(text.slice(i, i + 4), 16));
          i += 4;
        } else fail("PARSE_ERROR", `bad escape '\\${esc}'`);
      } else {
        out += ch;
        i++;
      }
    }
  }

  function parseNumber() {
    const start = i;
    if (text[i] === "-") i++;
    while (i < text.length && text[i] >= "0" && text[i] <= "9") i++;
    if (i === start || (text[start] === "-" && i === start + 1)) fail("PARSE_ERROR", "expected a value");
    const raw = text.slice(start, i);
    if (text[i] === "." || text[i] === "e" || text[i] === "E") {
      fail("PARSE_ERROR", "ACS-1 numbers are integers; floats are carried as strings");
    }
    const n = BigInt(raw);
    if (n < INT64_MIN || n > INT64_MAX) fail("INT_OUT_OF_RANGE", `outside signed 64-bit range: ${raw}`);
    return n;
  }

  const value = parseValue();
  if (i !== text.length) fail("PARSE_ERROR", "trailing content");
  return value;
}

export function isCanonical(bytes) {
  try {
    return Buffer.from(canonicalize(parseStrict(bytes))).equals(Buffer.from(bytes));
  } catch {
    return false;
  }
}
